File: week01/project-alpha/backend/app/schemas/ticket.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from app.schemas.label import LabelResponse


class TicketBase(BaseModel):
    title: str = Field(..., max_length=200, min_length=1)
    description: Optional[str] = Field(None, max_length=5000)


class TicketCreate(TicketBase):
    label_ids: Optional[List[int]] = []


class TicketUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=200, min_length=1)
    description: Optional[str] = Field(None, max_length=5000)


class TicketResponse(TicketBase):
    id: int
    status: str
    labels: List[LabelResponse] = []
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TicketListResponse(BaseModel):
    data: List[TicketResponse]
    pagination: dict


class TicketFilterParams(BaseModel):
    status: Optional[str] = None
    label_ids: Optional[List[int]] = None
    search: Optional[str] = None
    page: int = 1
    page_size: int = 20
    sort_by: str = "created_at"
    sort_order: str = "desc"

    class Config:
        from_attributes = True

File: week01/project-alpha/frontend/src/types/ticket.ts
export enum TicketStatus {
  OPEN = 'open',
  COMPLETED = 'completed',
  CANCELLED = 'cancelled',
}

export interface Label {
  id: number;
  name: string;
  color: string;
  created_at: string;
  updated_at: string;
}

export interface Ticket {
  id: number;
  title: string;
  description: string | null;
  status: TicketStatus;
  labels: Label[];
  created_at: string;
  updated_at: string;
}

export interface Pagination {
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

export interface TicketListResponse {
  data: Ticket[];
  pagination: Pagination;
}

export interface TicketFilter {
  status?: string;
  label_ids?: number[];
  search?: string;
  page?: number;
  page_size?: number;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

export interface CreateTicketRequest {
  title: string;
  description?: string;
  label_ids?: number[];
}

export interface UpdateTicketRequest {
  title?: string;
  description?: string;
}

export const STATUS_LABELS: Record<TicketStatus, string> = {
  [TicketStatus.OPEN]: 'Open',
  [TicketStatus.COMPLETED]: 'Completed',
  [TicketStatus.CANCELLED]: 'Cancelled',
};

export const STATUS_COLORS: Record<TicketStatus, string> = {
  [TicketStatus.OPEN]: 'bg-blue-100 text-blue-800',
  [TicketStatus.COMPLETED]: 'bg-green-100 text-green-800',
  [TicketStatus.CANCELLED]: 'bg-gray-100 text-gray-800',
};
